### ISSUE TITLE
VCH creation fails with "Failed to validate VCH: please specify a datacenter..."

### ISSUE THREAD

**Issue by lgayatri** [vmware/vic-ui#208]
@jooskim @mdubya66 

Build: vic-v1.3.0-rc3-2784-7d227d5c.ova

![image](https://user-images.githubusercontent.com/25006749/33764973-b6124220-dc3c-11e7-9f0b-0c7c761d5956.png)

If I copy the same Linux command  and run , I see an error as the target is mentioned incorrectly.

./vic-machine-linux create --name LG --compute-resource /Datacenter/host/Cluster --image-store CPBU_2TB_01 --base-image-size 8GB --bridge-network vxw-dvs-3444-virtualwire-1-sid-5000-VCH-LS-1 --bridge-network-range 172.16.0.0/12 --public-network DPortGroup --dns-server 10.142.7.21 --tls-cname LG --organization LG --certificate-key-size 2048 --user Administrator@VSPHERE.LOCAL --thumbprint C1:6D:3E:41:A5:89:5E:89:FB:DB:82:2F:32:82:8E:54:53:8B:65:34 --target wdc-stvc3944-146-dhcp15.eng.vmware.com --ops-user administrator@vsphere.local
INFO[0000] ### Installing VCH ####
INFO[0000] vSphere password for Administrator@VSPHERE.LOCAL:
INFO[0010] vSphere password for administrator@vsphere.local:
INFO[0015] Generating CA certificate/key pair - private key in LG/ca-key.pem
INFO[0016] Generating server certificate/key pair - private key in LG/server-key.pem
INFO[0016] Generating client certificate/key pair - private key in LG/key.pem
INFO[0017] Generated browser friendly PFX client certificate - certificate in LG/cert.pfx
INFO[0017] Validating supplied configuration
ERRO[0017] Datacenter must be specified in --target (e.g. https://addr/datacenter)
INFO[0017] Suggesting valid values for datacenter in --target
INFO[0017] Suggested datacenters:
INFO[0017]   "LGDatacenter"
INFO[0017]   "Datacenter"
ERRO[0017] Create cannot continue: configuration validation failed
ERRO[0017] --------------------
ERRO[0017] vic-machine-linux create failed: Datacenter must be specified in --target (e.g. https://addr/datacenter)

Issues here:
1. VCH creation through wizard should have passed
2. Command  should append https to target 

Setup:
VCSA: wdc-stvc3944-146-dhcp15.eng.vmware.com
administrator@vsphere.local/Admin!23
root/Admin!23

**Comment by jak-atx** [comment]
@zjs do you have any idea why vch creation would be failing here? I'm not quite clear from the errror message what the issue is.

**Comment by mlh78750** [comment]
Was the target selected from a list and then not formatted correctly for the API or the CLI? Is there a way for a customer to recover from this?

**Comment by zjs** [comment]
I don't know this codepath, but this makes it sound like we have some cases where the datacenter needs to be specified as a part of the target value:

> ```
> ERRO[0017] vic-machine-linux create failed: Datacenter must be specified in --target (e.g. https://addr/datacenter)
> ```

I'm _guessing_ that this is necessary when there are multiple datacenters.

When this is necessary, I guess we should:
1. Use the datacenter-specific API endpoint (`/target/.../datacenter/.../vch` instead of `/target/.../vch`).
2. Include the datacenter name in the `--target` argument when constructing the CLI examples.

> Is there a way for a customer to recover from this?

A work-around might be for the customer to copy the CLI example and add the datacenter to the  `--target` argument  themselves, but we should verify that would work. (Also, it's not a great work-around, since we haven't yet fixed #67, the quoting issue with the CLI examples.)

**Comment by jooskim** [comment]
Thanks for the detailed log. I believe appending the datacenter to the --target flag should fix the error you got from the CLI command. Will work on this fix

**Comment by lgayatri** [comment]
@mlh78750 , yes the target was selected from the UI's displayed list. As of now, there is not way out from UI.

**Comment by jak-atx** [comment]
@zjs is this endpoint (/target/.../datacenter/.../vch) one we can use regardless of if we have one or multiple datacenters?

**Comment by jak-atx** [comment]
@lgayatri can i get the IP of your vcsa?

**Comment by jak-atx** [comment]
@zjs I am seeing an odd issue, when I use the datacenter endpoint I get bad credential error:

https://10.192.102.207:8443/container/target/sc2-rdops-vm02-dhcp-93-240.eng.vmware.com/datacenter/datacenter-2/vch?thumbprint=8A:FD:1E:AF:EC:D3:31:46:FB:E3:9E:89:86:4D:D1:20:03:95:A2:F4

**Validation Error: Failed to log in to sc2-rdops-vm02-dhcp-93-240.eng.vmware.com: ServerFaultCode: Cannot complete login due to an incorrect user name or password.**

I use the exact same payload and credentials on this endpoint it goes through fine:

https://10.192.102.207:8443/container/target/sc2-rdops-vm02-dhcp-93-240.eng.vmware.com/vch?thumbprint=8A:FD:1E:AF:EC:D3:31:46:FB:E3:9E:89:86:4D:D1:20:03:95:A2:F4

payload:
{"name":"virtual-container-host-57575","compute":{"cpu":{"limit":{"units":"MHz","value":0}},"memory":{"limit":{"units":"MiB","value":0}},"resource":{"name":"/Datacenter/host/Cluster"}},"storage":{"image_stores":["datastore1"],"base_image_size":{"units":"GiB","value":8}},"network":{"bridge":{"ip_range":"172.16.0.0/12","port_group":{"name":"bridge"}},"public":{"port_group":{"name":"vm-network"}}},"auth":{"client":{"certificate_authorities":[]},"server":{"generate":{"size":{"value":2048,"units":"bit"},"organization":["virtual-container-host-57575"],"cname":"virtual-container-host-57575"}}},"container":{},"debug":0,"endpoint":{"operations_credentials":{"user":"administrator@vsphere.local","password":"Admin!23"}},"registry":{}}

**Comment by zjs** [comment]
I'm not sure what could be causing that. We have a variety of end-to-end tests which use the datacenter-scoped endpoints, but all our our end-to-end tests currently use basic authentication.

Presumably you're using session tickets? Is there any chance you're accidentally re-using a session ticket?

**Comment by jak-atx** [comment]
We are using the cloneSessionTicket, but I've logged in/out and restarted virgo server multiple times and still same issue. Not sure what else I can try.

**Comment by zjs** [comment]
Can you provide logs from vic-machine-sever?

I verified this command, which includes a datacenter and an ops user, succeeds against RC3:
```
curl -k -s -X POST https://$OVA_IP:8443/container/target/$VC_IP/datacenter/datacenter-41/vch?thumbprint=$THUMBPRINT -H "Authorization: Basic $AUTH" -H "Content-Type: application/json" --data '{"name":"vch-dc-test","compute":{"resource":{"name":"/dc1/host/cluster1/Resources"}},"storage":{"image_stores":["ds://sharedVmfs-1"]},"network":{"bridge":{"ip_range":"172.16.0.0/12","port_group":{"name":"network"}},"public":{"port_group":{"name":"PublicNetwork"}}},"endpoint":{"operations_credentials":{"user":"administrator@vsphere.local","password":"Admin!23"}},"auth":{"server":{"generate":{"cname":"vch.example.com","organization":["VMware, Inc."],"size":{"value":2048,"units":"bits"}}},"client":{"no_tls_verify": true}}}'
```

**Comment by cristianfalcone** [comment]
@zjs 

browser request:

```
POST /container/target/sc2-rdops-vm02-dhcp-93-240.eng.vmware.com/datacenter/datacenter-2/vch?thumbprint=8A:FD:1E:AF:EC:D3:31:46:FB:E3:9E:89:86:4D:D1:20:03:95:A2:F4 HTTP/1.1
Host: 10.192.102.207:8443
Connection: keep-alive
Content-Length: 756
Pragma: no-cache
Cache-Control: no-cache
Accept: application/json, text/plain, */*
Origin: https://localhost:9443
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36
X-VMWARE-TICKET: cst-VCT-520d7bbd-3788-81ba-78e1-173188528db4--tp-8A-FD-1E-AF-EC-D3-31-46-FB-E3-9E-89-86-4D-D1-20-03-95-A2-F4
Content-Type: application/json
Referer: https://localhost:9443/ui/vic/resources/build-dev/create-vch/urn:vic:vic:Root:vic%25252Fvic-root
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9,es;q=0.8
```

request body:

```json
{
   "name":"virtual-container-host-9",
   "compute":{
      "cpu":{
         "limit":{
            "units":"MHz",
            "value":0
         }
      },
      "memory":{
         "limit":{
            "units":"MiB",
            "value":0
         }
      },
      "resource":{
         "name":"/Datacenter/host/Cluster/10.192.111.11"
      }
   },
   "storage":{
      "image_stores":[
         "datastore1"
      ],
      "base_image_size":{
         "units":"GiB",
         "value":8
      }
   },
   "network":{
      "bridge":{
         "ip_range":"172.16.0.0/12",
         "port_group":{
            "name":"bridge"
         }
      },
      "public":{
         "port_group":{
            "name":"network"
         }
      }
   },
   "auth":{
      "client":{
         "certificate_authorities":[

         ]
      },
      "server":{
         "generate":{
            "size":{
               "value":2048,
               "units":"bit"
            },
            "organization":[
               "virtual-container-host-9"
            ],
            "cname":"virtual-container-host-9"
         }
      }
   },
   "container":{

   },
   "debug":0,
   "endpoint":{
      "operations_credentials":{
         "user":"administrator@vsphere.local",
         "password":"Admin!23",
         "grant_permissions":true
      }
   },
   "registry":{

   }
}
```

response headers:

```
HTTP/1.1 400 Bad Request
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: Content-Length
Content-Type: application/json
Vary: Origin
Date: Tue, 12 Dec 2017 16:12:44 GMT
Content-Length: 177
````

response body:

```json
{
   "message":"Validation Error: Failed to log in to sc2-rdops-vm02-dhcp-93-240.eng.vmware.com: ServerFaultCode: Cannot complete login due to an incorrect user name or password."
}
```

vic-machine-server log:

```
root@Photon [ /storage/log/vic-machine-server ]# tail -F vic-machine-server.log  
Dec 12 2017 15:19:09.134Z INFO  op=1.838: Response: 200 OK
Dec 12 2017 15:20:41.337Z INFO  op=1.839: Request: GET /container/hello HTTP/1.1
Dec 12 2017 15:20:41.337Z INFO  op=1.839: Response: 200 OK
Dec 12 2017 16:07:55.265Z INFO  op=1.840: Request: GET /container/hello HTTP/1.1
Dec 12 2017 16:07:55.266Z INFO  op=1.840: Response: 200 OK
Dec 12 2017 16:08:57.083Z INFO  op=1.841: Request: OPTIONS /container/target/sc2-rdops-vm02-dhcp-93-240.eng.vmware.com/datacenter/datacenter-2/vch HTTP/1.1
Dec 12 2017 16:08:57.083Z INFO  op=1.841: Response: 200 OK
Dec 12 2017 16:08:57.332Z INFO  op=1.842: Request: POST /container/target/sc2-rdops-vm02-dhcp-93-240.eng.vmware.com/datacenter/datacenter-2/vch HTTP/1.1
Dec 12 2017 16:08:57.394Z INFO  op=1.842: Validating target
Dec 12 2017 16:08:57.410Z INFO  op=1.842: Response: 400 Bad Request
```

I tested the same POST request without `/datacenter/datacenter-2` and it successfully created the VCH.

**Comment by cristianfalcone** [comment]
As a workaround, using the CLI command line with `--target sc2-rdops-vm02-dhcp-93240.eng.vmware.com/Datacenter` works just fine.

**Comment by zjs** [comment]
I've identified the root cause of the "Validation Error: Failed to log in to ..." and filed vmware/vic#6947 to track fixing it.

**Comment by lgayatri** [comment]
This issue is consistent. Hope to get the fix in next build.

![image](https://user-images.githubusercontent.com/25006749/33922164-d4b1e158-dfee-11e7-9721-6c10ad72b43c.png)

**Comment by lgayatri** [comment]
Still seeing this issue:
![image](https://user-images.githubusercontent.com/25006749/34067200-9ed2167c-e244-11e7-91a9-75f37f3c43e9.png)

**Comment by AngieCris** [comment]
@lgayatri Does this pr: https://github.com/vmware/vic-ui/issues/257 capture the same failure?
The `target validation failed` failure could have a different cause from the `please specify a datacenter` failure in this ticket

**Comment by lgayatri** [comment]
@AngieCris ,https://github.com/vmware/vic-ui/issues/257 is when we expand a VCH which was created with CLI. These are different issues.

**Comment by lgayatri** [comment]
I verified with RC4 and here is the latest repro of today morning

![image](https://user-images.githubusercontent.com/25006749/34285516-2e1a86ec-e700-11e7-8d09-04e3d99d67ad.png)

command that gets generated:
```
./vic-machine-linux create --name LG-Test\  --compute-resource /Datacenter/host/Cluster/w2-hs2-d2612.eng.vmware.com --cpu 1000 --memory 100 --image-store CPBU_2TB_01 --base-image-size 8GB --bridge-network vxw-dvs-3444-virtualwire-1-sid-5000-VCH-LS-1 --bridge-network-range 172.16.0.0/12 --public-network DPortGroup --tls-cname LG-Test\  --organization LG-Test\  --certificate-key-size 2048 --no-tlsverify --user Administrator@VSPHERE.LOCAL --thumbprint C1:6D:3E:41:A5:89:5E:89:FB:DB:82:2F:32:82:8E:54:53:8B:65:34 --target wdc-stvc3944-146-dhcp15.eng.vmware.com/Datacenter --ops-user Administrator@vsphere.local
```

**Comment by AngieCris** [comment]
@lgayatri I believe the above `Target validation failed: Datacenter "Datacenter" in --target is not found` error has a different root cause from the `Failed to validate VCH: please specify a datacenter...` error described in original issue. 
The root cause should be the same as https://github.com/vmware/vic-ui/issues/257, although they happen in different scenarios.

**Comment by mdubya66** [comment]
Closing this as fixed by vmware/vic#6994. If it is not we should re-open.
